k8s reporter, Extend collector to include network commands

oshoval opened the pull request:
Extend k8s reporter to collect networking commands upon assertion 
or when triggered by `make dump`.

Add commands for
1. Nodes (see `executeNodeCommands`)
2. Virt-launchers (see `executeVirtLauncherCommands`)
3. VMIs (see `executeVMICommands`)

This PR also fixes re-login of alpine.
 
Signed-off-by: Or Shoval <oshoval@redhat.com>

```release-note
None
```

alonSadan commented on tests/reporter/kubernetes.go:
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
+func getVmiType(vmi v12.VirtualMachineInstance) (string, error) {
+	for _, volume := range vmi.Spec.Volumes {
+		if volume.VolumeSource.ContainerDisk == nil {
+			continue
```
I think it would help with debugging if you will return a different error here
```
return "", fmt.Errorf("containerDisk is nil for vmi  %s", vmi.ObjectMeta.Name)
```

oshoval replied:
Some volumes don't have ContainerDisk and its ok, there might be couple volumes, we don't want to print all the non ContainerDisk ones, but we do `continue`

alonSadan commented on tests/reporter/kubernetes.go:
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
+		}
+	}
+
+	return "", fmt.Errorf("unknown type, vmi %s", vmi.ObjectMeta.Name)
```
If you choose to return if containerDisk is nil, you can also return the unknown type here:
```
return "", fmt.Errorf("unknown type %s for vmi %s", volume.VolumeSource.ContainerDisk.Image, vmi.ObjectMeta.Name) 
```

oshoval replied:
It might have several ContainerDisk volumes IIUC,
if so i can't print them

alonSadan commented on tests/reporter/kubernetes.go:
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
+	return err
+}
+
+func checkPodComputeContainer(pod v1.Pod) bool {
```
nit: maybe change to `isPodComputeContainer`

oshoval replied:
but the pod is not a compute container, it might have a compute container, this is what the function checks

alonSadan replied:
exactly, boolean functions like this one, are a lot of times called `isSomething`  or `areSomething`.
for exapmle check out the [first result](https://levelup.gitconnected.com/knot-of-nots-avoiding-negative-names-for-boolean-methods-641896a94a42#:~:text=The%20usual%20convention%20to%20name,exists%2C%20you%20would%20say%20user.) in google for "boolean function naming convention".

there is no known convention for `check Something` AFAIK, and it could mean many things.

alonSadan reviewed (commented):
Thanks Or. Really nice. some minor comments

oshoval commented:
Thanks Alon,
addressed comments

alonSadan commented on tests/reporter/kubernetes.go:
```diff
@@ -134,6 +142,10 @@ func (r *KubernetesReporter) Dump(duration time.Duration) {
 
 	r.logSRIOVInfo(virtCli)
+
+	r.logNodeCommands(virtCli)
```
Is there anything else other than network info that is retrieved from `logNodeCommands`, `logVirtLauncherCommands`, and `logVMICommands` ?
Perhaps it would be better to put it in a "network context" and call those functions :
`logNodeNetworkInfo`, `logVirtLauncherNetworkInfo`, and `logVMINetworkInfo`?
or, the same but with 'NetworkState` instead of `Commands`?

oshoval replied:
Answering on all the comments at once:
its almost true that they run only network commands but
1. dmesg is not a network command, so maybe not perfectly true.
2. we might want to add them to the future commands which are not network

having said that, it dumps the info in network folder

considered that, wdyt we should do ? (about all comments, since there are two groups ,this one, and the inner functions of the other comments)

alonSadan replied:
I see your point. Fine by me to leave it as is

alonSadan commented on tests/reporter/kubernetes.go:
```diff
@@ -337,6 +349,100 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
+				continue
+			}
+
+			r.executeVMICommands(vmi, logsdir, vmiType)
```
same as here https://github.com/kubevirt/kubevirt/pull/6116/files#r677171619
maybe, since it looks like this is only networking commands, rename this function to show this is only network related commands.

alonSadan commented on tests/reporter/kubernetes.go:
```diff
@@ -337,6 +349,100 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
+			continue
+		}
+
+		r.executeNodeCommands(virtCli, logsdir, pod)
```
ditto

alonSadan commented on tests/reporter/kubernetes.go:
```diff
@@ -337,6 +349,100 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
+			continue
+		}
+
+		r.executeVirtLauncherCommands(virtCli, logsdir, pod)
```
ditto

alonSadan reviewed (approved):
Thank you:)

oshoval commented:
Thanks Alon
@AlonaKaplan can you take a look please?

oshoval commented on tests/reporter/kubernetes.go:
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
+		{command: "bridge fdb", fileNameSuffix: "brfdb"},
+		{command: "lspci", fileNameSuffix: "lspci"},
+		{command: "env", fileNameSuffix: "env"},
+		{command: "nft list ruleset", fileNameSuffix: "nftlist"},
```
The binary is there, but output is empty indeed, i think youright
is there a simple code example please on how to access the launcher namespace from virt-handler ?

AlonaKaplan replied:
That's how it is [done](https://github.com/kubevirt/kubevirt/blob/65656ff9141ba830e03be5cbdf32439a943ea68a/pkg/virt-handler/isolation/detector.go#L77) in the virt-handler code. I guess you can simulate the same behaviour.
Another option is to do on the virt-handler -
`ps -ef | grep " --name vmi-masquerade " | grep /usr/bin/virt-launcher | head -n1 | cut -d" " -f7` to find the launcher pid and then enter the pid using nsenter.

oshoval commented on tests/reporter/kubernetes.go:
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
+}
+
+func prepareVmiConsole(vmi v12.VirtualMachineInstance, vmiType string) error {
+	err := console.SafeExpectBatch(&vmi, []expect.Batcher{
```
the login fails if its already logged
for example if the test did login already and then asserted

AlonaKaplan replied:
Our login methods should support re-login. If one of the three doesn't support it, it should be fixed.

oshoval replied:
Done

oshoval replied:
Note that the drawback of it (it was existing already on fedora and cirros, i just added to alpine)
is that it will consume several seconds in the tests themself for every first login

oshoval commented on tests/reporter/kubernetes.go:
```diff
@@ -337,6 +349,100 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
+
+	vmiNamespaces := tests.TestNamespaces
+	if isExternal {
+		vmiNamespaces = []string{v1.NamespaceAll}
```
Consoling a VM is disruptive, it disconnected the previous logged instances.
If we run in parallel and collect all VMs, we will hurt healthy tests during their session (or other ongoing assertions),

when running from e2e assertion, we can use `TestNamespaces` in order to touch only the namespaces
where tests should run their VMs, and each parallel instance of e2e test has already its own `TestNamespaces`

If we run `make dump` manually, we do want to collect all

AlonaKaplan commented on tests/reporter/kubernetes.go:
```diff
@@ -918,3 +1024,157 @@ func writeStringToFile(filePath string, data string) error {
+		{command: networkPrefix + "bridge fdb", fileNameSuffix: "brfdb"},
+		{command: networkPrefix + "nft list ruleset", fileNameSuffix: "nftlist"},
+
+		{command: hostPrefix + "/usr/bin/" + networkPrefix + "/usr/sbin/iptables --list", fileNameSuffix: "iptables"},
```
I think the interesting part in the iptables output is the counters. I would add the `-v` option.

oshoval commented:
addressed comments

oshoval commented:
we can see nftlist on kind here for example
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6116/pull-kubevirt-e2e-kind-1.19-sriov/1420712515519647744

(network/pods folder)

oshoval commented:
addressed comments

oshoval commented:
/retest 

doesnt seem related but lets make sure
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6116/pull-kubevirt-e2e-k8s-1.19-sig-storage/1422158509151096832

oshoval commented:
another weird failures
retesting but noting down 
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6116/pull-kubevirt-e2e-k8s-1.20-sig-compute-nonroot/1422158508349984768

/retest

AlonaKaplan commented on tests/reporter/kubernetes.go:
```diff
@@ -337,6 +350,133 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
+			continue
+		}
+
+		processList, err := getProcessList(virtCli, &virtHandlerPod)
```
Instead of the suggested code (that I believe is very complicated) I would create a map node_name -> pair list(virt_launcher_pod_name, vm_name).
Then for each virt-handler I would execute for each vmi `pgrep -f "name vmi_name.*no-fork"`, it should give you the pid of the virt-launcher pod.

What do you think?

oshoval replied:
pgrep didn't work well via console
we also need the uid, in order to find the virt-launcher name (in order to save to the right file name), so parsing is needed anyhow

side note: the right thing to do is use the uid instead of vmi_name + (namespace which is needed)
as we are doing here `GetRunningPodByLabel`

AlonaKaplan commented on tests/reporter/kubernetes.go:
```diff
@@ -337,6 +350,133 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
+		lines := strings.Split(processList, "\n")
+		for _, line := range lines {
+			if strings.Contains(line, "no-fork") {
+				ppid, uid, namespace := parseProcessLine(line)
```
Why the ppid is required isn't the pid enough. Both the for and the regular process have the same network name space.

oshoval replied:
pid can be used i believe as well indeed, just chosen the original process because it felt more "natual"
i will try the pid

AlonaKaplan commented on tests/console/login.go:
```diff
@@ -77,7 +77,20 @@ func LoginToAlpine(vmi *v1.VirtualMachineInstance) error {
 	b := append([]expect.Batcher{
+		&expect.BSnd{S: "\n"},
+		&expect.BExp{R: PromptExpression},
+		&expect.BSnd{S: "true\n"},
```
What is the added value of sending `true`?

oshoval replied:
same as https://github.com/kubevirt/kubevirt/pull/6116/files#r681193075

AlonaKaplan commented on tests/console/login.go:
```diff
@@ -117,10 +130,11 @@ func LoginToFedora(vmi *v1.VirtualMachineInstance) error {
 	b := append([]expect.Batcher{
 		&expect.BSnd{S: "\n"},
 		&expect.BExp{R: PromptExpression},
+		&expect.BSnd{S: "true\n"},
```
I don't understand why the `true` command is needed here. Couldn't understand the explanation in the commit message as well.

oshoval replied:
Lets say one sends command that fails, but its output isn't consumed 
then the relogin will find the PromptExpression but the echo $? will failed and the flow will continue the login process below (it will find the prompt there, so it won't send user:pass again, but it will reconfigure again the shell, and the whole thing is slower)

site node - this can be improved on other PR to match the full prompt (localhost|%s) login: `, vmi.Name)
instead of PromptExpression
because PromptExpression can be even matched during boot when the machine isn't actually logged
see here for example test 1705 when i totally removed the 
```
		&expect.BSnd{S: "echo $?\n"},
		&expect.BExp{R: RetValue("0")},
```
https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6116/pull-kubevirt-e2e-k8s-1.20-sig-network/1422158507934748672
Then it failed because it didn't need more condition after it matched the `# `
the `echo $?` makes it better but still not perfect but not need to be part of this PR

even adding the true can be done on other PR if needed basically, but it does improve the login in case the last command fail and its output was not consumed (for example timeout, or if ran manually and then ran `make dump`)

oshoval commented on tests/reporter/kubernetes.go:
```diff
@@ -918,3 +1069,240 @@ func writeStringToFile(filePath string, data string) error {
+	command := []string{
+		"/bin/bash",
+		"-c",
+		"trap '' URG && /usr/bin/ps -ef",
```
ps might panic due to URG signal

```
Signal 23 (URG) caught by ps (3.3.16).
/usr/bin/ps:ps/display.c:66: please report this bug
```

see here please
https://github.com/kubevirt/kubevirt/pull/5846/commits/3b16366b0af8f162701460cce57f9b0624358968

oshoval commented on tests/reporter/kubernetes.go:
```diff
@@ -337,6 +350,133 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
+			continue
+		}
+
+		lines := strings.Split(processList, "\n")
```
This is what i tried to do with `parseProcessLine` (line 437) and line (429), so please elaborate
lets discuss offline about all the points here if possible (this one and https://github.com/kubevirt/kubevirt/pull/6116/files#r681184802)
since they are related to each other and have many small details

oshoval commented:
addressed the comments that about the collector (not yet about the login commit)

oshoval commented:
Improved the login
Please see 
https://github.com/kubevirt/kubevirt/compare/a42be5ebf59b64c9298ab94599eb078a1174b640..bbc94be34e5075fe1cae400e9cd80078a0cc8da4

We might want to do "sudo su" again on the relogin block ?
(in case a user connected but without sudo)
currently the commands that we are collecting on the VM doesn't need sudo, so currently it 
shouldn't be an issue

oshoval commented:
Removed function i dont need anymore

AlonaKaplan commented on tests/reporter/kubernetes.go:
```diff
@@ -337,6 +352,142 @@ func (r *KubernetesReporter) logAuditLogs(virtCli kubecli.KubevirtClient, logsdi
+	}
+
+	vmis, err := virtCli.VirtualMachineInstance(v1.NamespaceAll).List(&metav1.ListOptions{})
+	if err != nil {
```
It looks great overall, just a small question. Maybe it is enough to query for the list of virt-launcher pods and no need to query the vmi list. Since the virt-launcher pod contains the name of the vmi?

oshoval replied:
I think you are right (you mean the name of the virt-launcher i guess which is what we need)
trying

AlonaKaplan commented on tests/console/login.go:
```diff
@@ -116,11 +126,9 @@ func LoginToFedora(vmi *v1.VirtualMachineInstance) error {
-		&expect.BExp{R: PromptExpression},
-		&expect.BSnd{S: "echo $?\n"},
-		&expect.BExp{R: RetValue("0")},
+		&expect.BExp{R: fmt.Sprintf(`(\[fedora@(localhost|%s) ~\]\$ |\[root@(localhost|%s) fedora\]\# )`, vmi.Name, vmi.Name)},
```
I guess cirros needs a similar fix. Unless it blocks this PR I believe the login part deserves a separate PR.

oshoval replied:
https://github.com/kubevirt/kubevirt/pull/6181

oshoval replied:
On Cirros there is no issue, because the prompt of both sudo and non sudo are just `$` and this is what it waits

if `$` was printed in the middle of the boot it was an issue, but then we would saw it before already

oshoval commented:
Rebased

oshoval commented:
Addressed comments
added this as well
https://github.com/kubevirt/kubevirt/compare/b4a7fbd35ed63260c2da5190a43e7f1d91c0eb88..9347fe3db7298502f02f1a15e085ddf1c8032944#diff-dc90a95f97ee6623eb2cd8b29fd07ed28cda2fa3cc87aa75127900ec76da5142R80

AlonaKaplan commented on tests/console/login.go:
```diff
@@ -77,7 +77,22 @@ func LoginToAlpine(vmi *v1.VirtualMachineInstance) error {
 	}
 	defer expecter.Close()
 
+	err = expecter.Send("\n")
```
Why lines 80-84 are needed? isn't the batcher enough?

oshoval replied:
Its basically enough, but those are little faster, sanity that we can even send,
saw it as well on fedora and Cirros so added here

AlonaKaplan commented:
/approve Thanks!

oshoval commented:
Thanks Alona
@alonSadan can you take a look please?

oshoval commented:
Thanks Alon

/retest
